components/feature/syncedtabs/src/main/java/mozilla/components/feature/syncedtabs/internal/SyncedTabsController.kt

```diff
@@ -0,0 +1,48 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.syncedtabs.internal
+
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import mozilla.components.feature.syncedtabs.SyncedTabsStorage
+import mozilla.components.feature.syncedtabs.view.SyncedTabsView
+import mozilla.components.feature.syncedtabs.view.SyncedTabsView.ErrorType
+import mozilla.components.service.fxa.manager.FxaAccountManager
+import mozilla.components.service.fxa.manager.ext.withConstellation
+import kotlin.coroutines.CoroutineContext
+
+internal class SyncedTabsController(
+    private val storage: SyncedTabsStorage,
+    private val accountManager: FxaAccountManager,
+    private val view: SyncedTabsView,
+    private val coroutineContext: CoroutineContext
+) {
+
+    fun syncTabs() {
+        view.startLoading()
+
+        val scope = CoroutineScope(coroutineContext)
+
+        scope.launch {
+            accountManager.withConstellation {
+                val syncedTabs = storage.getSyncedTabs()
+                val otherDevices = state()?.otherDevices
+
+                scope.launch(Dispatchers.Main) {
+                    if (syncedTabs.isEmpty() && otherDevices?.isEmpty() == true) {
+                        view.onError(ErrorType.MULTIPLE_DEVICES_UNAVAILABLE)
+                    } else {
+                        view.displaySyncedTabs(syncedTabs)
+                    }
+                }
+            }
+
+            scope.launch(Dispatchers.Main) {
+                view.stopLoading()
+            }
+        }
+    }
```

Comment:
What happens here if we're in the "Connected to Sync, but tabs haven't been synced yet" state? Should we not call the `stopLoading()` method then?

Reply:
We call `view.stopLoading()` always at the end. Is there something else we should do?

Reply:
I'm concerned that we would show an empty list instead of a loading screen here, maybe we shouldn't call stopLoading yet?

Reply:
The `storage.getSyncedTabs()` is the async API that is invoked within the same coroutine scope that we invoke `view.stopLoading()` to make sure those happen synchronously.

So we would stay on the `startLoading` screen until that entire scope completes.